### ISSUE TITLE
Manually trigger an update for a dependency

### ISSUE THREAD

**Issue by 0xdevalias** [dependabot/dependabot-core#2980]
Currently dependabot runs on a schedule, and by 'some magic' decides which of the outdated dependencies it will open a PR to update next. We tend to keep this limit to a relatively small number of PR's (2), and often need to take some extra time to ensure dependency changes don't affect our codebase's stability (not nearly enough test coverage to trust that in an automated way yet)

While obviously we can update dependencies manually when needed (which is our current workaround), we've grown quite fond of the way dependabot captures all of the relevant changelog/etc information into the PR when it opens one.

It would be great if we were able to 'trigger' dependabot in some way to automatically bump a specific outdated dependency (probably with the ability to bypass the 'max open PRs' limit for these 'manual bumps'). The only 'difference' here would be getting triggered manually, it would still open the PR and do everything else in the same way as it currently does.

**Comment by thepwagner** [comment]
👋 thanks for the cool idea!

Today you can trigger Dependabot manually in the "Insights -> Dependency Graph -> Dependabot" section of your repository, but as noted this can't be targeted to a single dependency and is affected by the open PR limit.

**Comment by Mogost** [comment]
Seems that manually trigger does not work for `package-ecosystem: "github-actions"`.

**Comment by websolutions-hamburg** [comment]
Manually trigger for `package-ecosystem: "github-actions"` works for me.

**Comment by chenrui333** [comment]
I am also interested about this feature as well. :)

**Comment by techieshark** [comment]
Curious if anyone has a workaround handy. 

My current use case is to split code across two repos, and have a version change in repo A trigger a PR to update the lockfile in repo B (which relies on repo A). I could make a GitHub Action on repo A to run `npm`/`composer` in repo B, then automatically open a PR, but if I can avoid writing that code that'd be great (esp. if I use this pattern in various projects).

**Comment by thepwagner** [comment]
@techieshark you can push the updates to the `.github/dependabot.yaml` in order to trigger version updates immediately: https://docs.github.com/en/code-security/supply-chain-security/keeping-your-dependencies-updated-automatically/configuration-options-for-dependency-updates

Having repository A push the configuration file in repository B will update _all_ dependencies, including repository A. It is wasteful, but the best workaround I know of atm.

**Comment by mwaddell** [comment]
Have you tried having a github action which runs when a dependabot-created PR is closed and issues a POST request to `/{owner}/{repo}/network/updates?update_config_id={update_config_id}` to force a new dependabot run?

**Comment by elgohr** [comment]
@mwaddell would need to do web-scraping to get the cookies (as the url is not an API-URL), right?

**Comment by mwaddell** [comment]
> @mwaddell would need to do web-scraping to get the cookies (as the url is not an API-URL), right?

Yes, you're correct, so it's only useful for doing a manual trigger from a script (instead of having to go find the correct button through the UI), but it's not going to work if you're trying to script it from github actions.

If you want to trigger dependabot to rerun through github actions, the only ways I know of are pushing a change to the dependabot.yaml file (not a great solution) or adding comments to specific pull requests that dependabot has created.

**Comment by weronikadominiak** [comment]
Hi!
I am also interested in being able to trigger dependabot re-run from Github Actions. Any chance of having that in future releases? 

Thanks!

**Comment by SalimBensiali** [comment]
I would love to be able to do this too. A use case for this is when dependabot raises a security alert for a dependency, but is unable to raise a PR due to version conflicts. This usually happens when one or more dependencies are blocking a security update on a child dependency. When this occurs, what I am thinking to programatically be able to achieve in a workflow is:
- Find the dependencies metadata preventing the security update by parsing the dependabot PR error status (from the relevant GraphQL endpoint)
- For each of the said dependencies, instruct dependabot to raise a version update PR even if the of open pull request limit is exceeded. This is because it is already the case for security update PRs, which this particular type of version update falls under. I.e. it is a "regular" dependabot version update which unblock a dependabot security update.

What would be even better is iff this sort of behaviour was baked into dependabot by default.

**Comment by lestephane** [comment]
Until GitHub starts tracking dependents of private repositories, and use that information to trigger the smallest set of dependabot update checks, people will be forced to come up with DIY (and wasteful) solutions.

This needs to happen at least when the dependents belong to the same owner / org / enterprise.

**Comment by SalimBensiali** [comment]
The use case I mentioned above seems to be rendered completely obsolete by this recent update to Dependabot https://github.blog/changelog/2022-09-07-dependabot-unlocks-transitive-dependencies-for-npm-projects/ 🎉

**Comment by big-andy-coates** [comment]
> Have you tried having a github action which runs when a dependabot-created PR is closed and issues a POST request to `/{owner}/{repo}/network/updates?update_config_id={update_config_id}` to force a new dependabot run?
> 
> 

The limiting factor here is obviously the non-deterministic `update_config_id`.

Would it be possible to make this deterministic?

Maybe accepting an `id` field in the dependabot yml file?

If it's a deterministic url, then it would be easy enough for community members to create GitHub actions, etc.

**Comment by jeffwidman** [comment]
Related:
* https://github.com/dependabot/dependabot-core/issues/2201

Although I do wonder if some of the need for this would disappear if we added support for realtime updates:
* https://github.com/dependabot/dependabot-core/issues/3488

**Comment by amacneil** [comment]
It would be very helpful if Github could offer a REST API endpoint to trigger a dependabot run, so that we don't need to use web scraping / stored cookies.

Looks like this issue is in the top 20 most 👍 on dependabot, and would be a quick win since there is already clearly a backend API for this.

**Comment by paulz** [comment]
is there a way to test if we configured dependabot correctly?

**Comment by jeffwidman** [comment]
@paulz please file a new issue for a question like that rather than posting on a somewhat unrelated thread, it creates noise for everyone else watching this issue for updates.

I moved your comment to #7125

**Comment by duaneking** [comment]
Dependabot is utterly broken.

It's not taking requests.

It's not giving errors.

There is no way to debug it.

We have no way to know what it's doing, or when it will respond.

At this point, it's become a distraction for me and my team. 

We just want to be able to tell the robot what to do so we can have it do the right thing, instead of it acting stupid and blocking builds and wasting team members time and actively getting in the way.

This actively makes Github a pain to use, and creates problems for teams because now we're dependent on a third party entity in the form of this AI robot thing that we can't control, that actively hurts our productivity, and acts like the sort of employee we would normally never want to work with and would never hire because its so Irrationally unreliable that we don't want the bot on our team if we can't control it and make sure it acts safely.

To be clear, the current incarnation of dependabot is a bad actor, and we can't tell it apart from a supply chain attack at the moment, Because it's so unreliable.

Virtual or human, if somebody wants to be of assistance to the team, they need to be able to take direction and fix things, and thats an expectation that every person has for dependabot, And that expectation is being failed.

**Comment by mwaddell** [comment]
@duaneking Unfortunately, I have to agree with you.  I started using Dependabot before GitHub bought it because it automated a lot of the busywork parts of managing software security and compliance.  However, the project has been largely stagnant since then.  The internal Dependabot team has been focused exclusively on scaling the existing code, and there has been little to no support for new features nor any coordinated governance of the open source community that cropped up around Dependabot to allow contributors to add these features ourselves.  On top of that, major, critical portions of Dependabot are closed source, so there are major gaps in what the open source community can even help with had there been good community governance in place.

I had hoped that GitHub would model better open source project governance on this project, but (as evidenced by the hundreds of years-old issues that have remained untouched), that has not proven to be the case.  In the meantime, I (and many others) have left for https://github.com/renovatebot/renovate which is vastly more configurable and already supports most of the feature requests that have been sitting open in this project for years.  It does not support the same level of integration into GitHub as Dependabot, but it more than makes up for that with its level of configurability and true openness.

**Comment by duaneking** [comment]
Unfortunately, any claim that they're simply working on scaling the code would require that they actually do the QA to make sure that basic things that we're already working don't break. And since they're clearly not doing that, since things are clearly broken, it seems to me like that team is simply sitting on its hands and ignoring the community.

**Comment by htpaf** [comment]
@mwaddell 
I was looking at https://github.com/dependabot/cli as by the name alone it seemed that it could work to "trigger" dependabot just like GitHub `cli `can trigger workflows.

It does not seem to be used for more than testing but could some small modification to the` dependabot/cli `be used as workaround perhaps, what do you think?

**Comment by LeftTwixWand** [comment]
Are there any updates on this topic?

**Comment by mwaddell** [comment]
> @mwaddell I was looking at https://github.com/dependabot/cli as by the name alone it seemed that it could work to "trigger" dependabot just like GitHub `cli `can trigger workflows.
> 
> It does not seem to be used for more than testing but could some small modification to the`dependabot/cli`be used as workaround perhaps, what do you think?

No, dependabot/cli is used for local testing and does not actually connect to the "real" dependabot instance so it cannot create PRs, it just runs a local test version and outputs what the PR would be if it could create it.

**Comment by JoernBerkefeld** [comment]
> 👋 thanks for the cool idea!
> 
> Today you can trigger Dependabot manually in the "Insights -> Dependency Graph -> Dependabot" section of your repository, but as noted this can't be targeted to a single dependency and is affected by the open PR limit.

this no longer works though, doesn't it?

**Comment by TheJulianJES** [comment]
@JoernBerkefeld It does, but you also need to click on "Recent update jobs" -> "Check for updates" on the Dependabot page.

**Comment by andrewscfc** [comment]
dependabot recreate

**Comment by JonasKs** [comment]
A common use-case is to update k8s manifests whenever a new image is pushed to a registry, e.g. when you're making applications and want to push a new version to production, you could typically have this workflow:

- Merge feature/fix into `main`
- Create a new tag for the new release (which can be automated through e.g. [release-please](https://github.com/googleapis/release-please))
- Tag triggers a build-action, which builds a new docker container
**- When a new docker container is built, update kubernetes manifests to use new SHA**
- Sync k8s manifest changes through Argo CD

Dependabot works great for the step marked in **bold**, and updating the manifest file is simple with this `dependabot.yml`: 
```yaml
version: 2
registries:
  gh-registry:
    type: docker-registry
    url: https://ghcr.io
    username: jonasks
    password: ${{ secrets.GHCR_TOKEN }}
updates:
  - package-ecosystem: "docker"
    directory: "/deployment"
    registries: 
      - gh-registry
    schedule: 
      interval: "daily"
```
How ever, since it only runs on an interval, it could take up to 24h for the PR to appear - which would be about 23h59m too slow. 

I'd love if we could programmatically trigger dependabot, or add triggers (such as a new image to ghcr) instead of a daily schedule. 

The workaround is to update the manifests using `kustomize`+manually creating PRs in a workflow, `renovate` or `Argo CD Image Updater`.

**Comment by martin-morin** [comment]
When developing an application that relies on various repositories, it is often necessary to initiate integrations in other repositories. To streamline this process, we should be able to trigger Dependabot from an alternative workflow.

Currently, the existing constraints can lead to delays of several days before packages are fully integrated up to the top of the dependency pyramid.

The ability to initiate Dependabot when a workflow in another repository concludes is essential.

**Comment by mietzen** [comment]
> If you want to trigger dependabot to rerun through github actions, the only ways I know of are pushing a change to the dependabot.yaml file (not a great solution) or adding comments to specific pull requests that dependabot has created.

That actually helped! It's ugly but this works:

```sh
file=".github/dependabot.yml"

if [ -x "$file" ]; then
    chmod -x "$file"
else
    chmod +x "$file"
fi

git add "$file"
git commit -m "chore: trigger dependabot check"
git push
```

I going to create a github action for this, which I can trigger.
The downside is that we are going to spam out git log, since this only triggers when we push to the default branch -.- 

We could do something like:

```sh
git commit -m "chore: trigger dependabot check"
git push
git reset --hard HEAD~1
git push -f
```

But this just makes it even uglier and slightly risky to lose changes... A stable API endpoint would be some much better.

---

**Edit:** 

Here is my workflow, using a github app with `contents write` permission:

```yml
name: Trigger Dependabot

on:
  workflow_dispatch:

jobs:
  trigger-dependabot:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/create-github-app-token@v2
        id: app-token
        with:
          app-id: ${{ secrets.DEPENDABOT_TRIGGER_APP_ID }}
          private-key: ${{ secrets.DEPENDABOT_TRIGGER_APP_PRIVATE_KEY }}
          permission-contents: write

      - name: Get GitHub App User ID
        id: get-user-id
        run: echo "user-id=$(gh api "/users/${{ steps.app-token.outputs.app-slug }}[bot]" --jq .id)" >> "$GITHUB_OUTPUT"
        env:
          GH_TOKEN: ${{ steps.app-token.outputs.token }}

      - name: Set Git Committer
        run: |
          git config --global user.name '${{ steps.app-token.outputs.app-slug }}[bot]'
          git config --global user.email '${{ steps.get-user-id.outputs.user-id }}+${{ steps.app-token.outputs.app-slug }}[bot]@users.noreply.github.com'

      - name: Checkout Code
        uses: actions/checkout@v6
        with:
          token: ${{ steps.app-token.outputs.token }}
          fetch-depth: 2

      - name: Change dependabot config to trigger update
        run: |
          file=".github/dependabot.yml"

          # Check current mode and toggle
          if [ -x "$file" ]; then
              chmod -x "$file"
          else
              chmod +x "$file"
          fi
          git add "$file"
          git commit -m "chore: trigger dependabot check"
          git push
          # git reset --hard HEAD~1
          # git push -f
```

If you got branch protection enabled, this will enable you to add the GitHub to the bypass list, allowing it to push into the default branch / force push

This using `git push -f` will trigger dependabot twice, but so it is.

--- 

**Edit:**

To trigger the Upstream job from a Repo with our libs I use, our app needs the `actions write` permission:

```yaml
name: Trigger upstream Dependabot 

on:
  workflow_dispatch:
  release:
    types: [published]

jobs:
  trigger-dependabot:
    runs-on: ubuntu-24.04
    steps:
      - uses: actions/create-github-app-token@v2
        id: generate-token
        with:
          app-id: ${{ secrets.WORKFLOW_TRIGGER_APP_ID }}
          private-key: ${{ secrets.WORKFLOW_TRIGGER_APP_PRIVATE_KEY }}
          repositories: ${{ vars.UPSTREAM_REPO }} 
          permission-actions: write
      - name: Trigger upstream workflows
        uses: actions/github-script@v6
        with:
          github-token: ${{ steps.generate-token.outputs.token }}
          script: |
            await github.rest.actions.createWorkflowDispatch({
              owner: ${{ vars.UPSTREAM_REPO_OWNER }},
              repo: ${{ vars.UPSTREAM_REPO }},
              workflow_id: 'dependabot-trigger.yml',
              ref: 'master'
            })
```

---

You can also do both of these with a fine grained token.
